core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java

```diff
@@ -71,58 +64,56 @@ public ListBucketResult() {}
    * Creates an {@link ListBucketResult}.
    *
    * @param bucketName the bucket name
-   * @param objectsList a list of {@link URIStatus}, representing the objects
+   * @param children a list of {@link URIStatus}, representing the objects and common prefixes
    * @param options the list bucket options
    */
   public ListBucketResult(
-      String bucketName, List<URIStatus> objectsList, ListBucketOptions options) {
+      String bucketName, List<URIStatus> children, ListBucketOptions options) {
     mName = bucketName;
+    mMaxKeys = DEFAULT_MAX_KEYS;
+
     mPrefix = options.getPrefix();
-    mKeyCount = 0;
-    if (options.getMaxKeys() != null) {
-      mMaxKeys = Integer.parseInt(options.getMaxKeys());
-    }
-    mContents = new ArrayList<>();
-    mContinuationToken = options.getContinuationToken();
-
-    Collections.sort(objectsList, new URIStatusComparator());
-
-    int startIndex = 0;
-    if (options.getContinuationToken() != null) {
-      URIStatus tokenStatus = new URIStatus(new FileInfo().setPath(
-          mName + AlluxioURI.SEPARATOR + mContinuationToken));
-      startIndex = Collections.binarySearch(objectsList, tokenStatus, new URIStatusComparator());
-      if (startIndex < 0) {
-        // If continuation token does not exist in the object list, find the first element which is
-        // greater than the token.
-        startIndex = (-1) * startIndex - 1;
-      }
+    mMarker = options.getMarker();
+
+    Collections.sort(children, new URIStatusComparator());
+
+    final List<URIStatus> keys = children.stream()
+        .filter((status) -> status.getPath().compareTo(mMarker) > 0)
+        .limit(mMaxKeys)
+        .collect(Collectors.toList());
+
+    mKeyCount = keys.size();
+    mIsTruncated = mKeyCount == mMaxKeys;
+
+    if (mIsTruncated) {
+      mNextMarker = keys.get(keys.size() - 1).getPath();
```

Comment:
is it ever possible that the `keys.size() == 0`?

Reply:
not if mIsTruncated is True unless mMaxKeys is 0.